src/ucp/rndv/rndv_rkey_ptr.c

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "proto_rndv.inl"
+
+
+enum {
+    UCP_PROTO_RNDV_RKEY_PTR_STAGE_FETCH = UCP_PROTO_STAGE_START,
+    UCP_PROTO_RNDV_RKEY_PTR_STAGE_ATS
+};
+
+static ucs_status_t
+ucp_proto_rndv_rkey_ptr_init(const ucp_proto_init_params_t *init_params)
+{
+    ucp_context_t *context                = init_params->worker->context;
+    uint64_t rndv_modes                   = UCS_BIT(UCP_RNDV_MODE_RKEY_PTR);
+    ucp_proto_single_init_params_t params = {
+        .super.super         = *init_params,
+        .super.cfg_thresh    = ucp_proto_rndv_cfg_thresh(context, rndv_modes),
+        .super.cfg_priority  = 0,
+        .super.overhead      = 0,
+        .super.latency       = 0,
+        .super.min_length    = 0,
+        .super.max_length    = SIZE_MAX,
+        .super.min_frag_offs = UCP_PROTO_COMMON_OFFSET_INVALID,
+        .super.max_frag_offs = UCP_PROTO_COMMON_OFFSET_INVALID,
+        .super.max_iov_offs  = UCP_PROTO_COMMON_OFFSET_INVALID,
+        .super.hdr_size      = 0,
+        .super.memtype_op    = UCT_EP_OP_LAST,
+        .super.flags         = UCP_PROTO_COMMON_INIT_FLAG_RKEY_PTR |
+                               UCP_PROTO_COMMON_INIT_FLAG_RECV_ZCOPY |
+                               UCP_PROTO_COMMON_INIT_FLAG_REMOTE_ACCESS |
+                               UCP_PROTO_COMMON_INIT_FLAG_SINGLE_FRAG,
+        .lane_type           = UCP_LANE_TYPE_RKEY_PTR,
+        .tl_cap_flags        = 0,
+    };
+
+    if ((init_params->select_param->op_id != UCP_OP_ID_RNDV_RECV) ||
+        (init_params->select_param->dt_class != UCP_DATATYPE_CONTIG)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    return ucp_proto_single_init(&params);
+}
+
+static unsigned ucp_proto_rndv_progress_rkey_ptr(void *arg)
+{
+    ucp_worker_h worker = (ucp_worker_h)arg;
+    ucp_request_t *req  = ucs_queue_head_elem_non_empty(&worker->rkey_ptr_reqs,
+                                                        ucp_request_t,
+                                                        send.rndv.rkey_ptr.queue_elem);
+    size_t seg_size     = worker->context->config.ext.rkey_ptr_seg_size;
+    size_t length       = req->send.state.dt_iter.length;
+    size_t offset       = req->send.rndv.offset;
+    ucs_status_t status;
+    size_t new_offset;
+    void *src;
+    int last;
+
+    seg_size   = ucs_min(seg_size, length - offset);
+    new_offset = offset + seg_size;
+    last       = new_offset == length;
+    src        = UCS_PTR_BYTE_OFFSET(req->send.rndv.rkey_ptr.ptr, offset);
+    status     = ucp_datatype_iter_unpack(&req->send.state.dt_iter, worker,
+                                          seg_size, offset, src);
+    req->send.rndv.offset = req->send.state.dt_iter.offset = new_offset;
```

Comment:
req->send.state.dt_iter.offset should not be used; should use req->send.state.completed_size